### PR TITLE
use connector.findOrCreate() if possible

### PR DESCRIPTION
**Pull request by clarkorz**
When using mongodb as database, I often use the native `findAndModify()` to perform a `find or create` (using `$setOnInsert` and `upsert` option). Since `findAndModify()` is an atomic operation, it won't cause conflict in some cases.

This PR tests that if the `connector` support `findOrCreate`, and if so then use it, otherwise `find()` or `create()`.

**Note**:
Since this is an atomic operation now, should the `create` hooks be triggered?

Signed-off-by: Clark Wang <clark.wangs@gmail.com>

### CODE REVIEW

File: lib/connectors/memory.js [hunk 1]
```diff
@@ -228,6 +228,15 @@ Memory.prototype.updateOrCreate = function (model, data, callback) {
...
 
+Memory.prototype.findOrCreate = function (model, query, data, callback) {
+  this.all(model, query, function (err, list) {
+    if (err || (list && list[0])) return callback(err, list && list[0], false);
+    this.create(model, data, function (err) {
+      callback(err, data, true);
+    });
+  }.bind(this));
+};
+
```
Comment (bajtos) [inline lib/connectors/memory.js]: Isn't is better to leave this out and let juggler to use the default unoptimized implementation for the memory connector?

File: lib/dao.js [hunk 2]
```diff
@@ -332,14 +332,64 @@ DataAccessObject.findOrCreate = function findOrCreate(query, data, callback) {
...
+    data = obj.toObject(true);
+
+    // validation required
+    obj.isValid(function (valid) {
+      if (valid) {
+        findOrCreate();
+      } else {
+        callback(new ValidationError(obj), obj);
+      }
+    }, data);
```
Comment (bajtos) [inline lib/dao.js]: This looks like a lot of copy-and-pasted code, is there a way how to refactor it into a shared method?

**Comment by bajtos** [comment]
@raymondfeng could you please review ASAP? This patch affects my work on the new hooks infrastructure in https://github.com/strongloop/loopback-datasource-juggler/issues/367.

> Note: Since this is an atomic operation now, should the create hooks be triggered?

I'd say we can leave that out of this patch in favour of #367

File: lib/connectors/memory.js (hunk 1 above)
Comment (clarkorz) [inline lib/connectors/memory.js]: @bajtos you are right. I'm lazy. This only for testing the case that connector *has* `findOrCreate()`. But now, the tests can't cover the case that connector does NOT have `findOrCreate()`.

I should extend the memory connector and implement `findOrCreate()` in a standalone test.

Comment (bajtos) [inline lib/connectors/memory.js]: > I should extend the memory connector and implement findOrCreate() in a standalone test.

That would be much better indeed :+1:

**Comment by bajtos** [comment]
@clarkorz please rebase the patch on top of the current master. #403 modified the way how `findOrCreate` works, it added new hooks ([preliminary docs](https://gist.github.com/bajtos/b8157112452eeaf984c9#file-new-hooks-docs-md)).

**Comment by raymondfeng** [comment]
@clarkorz Any progress?

**Comment by clarkorz** [comment]
Ops...sorry for mess it up. I have dropped this PR and create a new one here https://github.com/strongloop/loopback-datasource-juggler/pull/419

**Comment by bajtos** [comment]
Please try to cleanup (rewrite) the git history using `git rebase -i` and/or `git cherry-pick` and keep the same PR next time.

**Pull request closed without merging**